### PR TITLE
FIX Empty relations don't have extra DB calls with eager-loading

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/10870

### CODE REVIEW

File: src/ORM/DataList.php [hunk 1]
```diff
@@ -1101,7 +1107,7 @@ private function fetchEagerLoadRelations(Query $query): void
                             $relationDataClass,
                             implode('.', $chainToDate),
                             $relationName,
-                            $parentRelationName
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1104]: The `$parentRelationName` argument was never used - in fact there wasn't even a parameter for it

File: src/ORM/DataList.php [hunk 2]
```diff
@@ -1367,6 +1420,46 @@ private function addEagerLoadedDataToParent(
...
+        string $relationDataClass,
+        ?string $relationListClass = null,
+        ?array $component = null
+    ): void {
+        foreach ($parentIDs as $id) {
+            // Build the empty list or record
+            switch ($relationType) {
+                case 'has_one':
+                    $dummyData = Injector::inst()->create($relationDataClass);
+                    break;
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1439]: `has_one` isn't used currently (see the comment about why it's not necessary in `fetchEagerLoadHasOne()`), but it's here for the sake of completion. It _is_ technically valid to pass in here, and shouldn't throw an exception if we do for some reason pass it in.

File: src/ORM/DataList.php [hunk 3]
```diff
@@ -1179,33 +1185,53 @@ private function fetchEagerLoadHasOne(
...
+        string $relationName,
+        string $relationType
     ): array {
+        $belongsToIDField = $component['joinField'];
         // Get ALL of the items for this relation up front, for ALL of the parents
         // Fetched as an array to avoid sporadic additional queries when the DataList is looped directly
         $fetchedRecords = DataObject::get($relationDataClass)->filter([$belongsToIDField => $parentIDs])->toArray();
         $fetchedIDs = [];
-
+        $foundParents = [];
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
        $foundParentIDs = [];
```

File: src/ORM/DataList.php [hunk 4]
```diff
@@ -1179,33 +1185,53 @@ private function fetchEagerLoadHasOne(
...
 
         // Add fetched record to the correct place
         foreach ($fetchedRecords as $fetched) {
             $fetchedIDs[] = $fetched->ID;
             $parentID = $fetched->$belongsToIDField;
-            $this->addEagerLoadedDataToParent($parents, $parentID, $relationChain, $relationName, $fetched, 'has_one');
-        }
+            $foundParents[] = $parentID;
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
            $foundParentIDs[] = $parentID;
```

File: src/ORM/DataList.php [hunk 5]
```diff
@@ -1179,33 +1185,53 @@ private function fetchEagerLoadHasOne(
...
+            $this->addEagerLoadedDataToParent($parents, $parentID, $relationChain, $relationName, $fetched, $relationType);
+        }
+
+        // Load empty DataObject records into any parents which have no child records
+        $emptyRelations = array_diff($parentIDs, $foundParents);
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
        $missingParentIDs = array_diff($parentIDs, $foundParentIDs);
```

File: src/ORM/DataList.php [hunk 6]
```diff
@@ -1179,33 +1185,53 @@ private function fetchEagerLoadHasOne(
...
+        $this->fillEmptyEagerLoadedRelations(
+            $parents,
+            $emptyRelations,
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
            $missingParentIDs,
```

File: src/ORM/DataList.php [hunk 7]
```diff
@@ -1234,12 +1261,24 @@ private function fetchEagerLoadHasMany(
...
                 $eagerLoadedLists[$parentID] = $eagerLoadedList;
-                $this->addEagerLoadedDataToParent($parents, $parentID, $relationChain, $relationName, $eagerLoadedList, 'has_many');
+                $this->addEagerLoadedDataToParent($parents, $parentID, $relationChain, $relationName, $eagerLoadedList, $relationType);
             }
             // Add this row to the list
             $eagerLoadedList->addRow($row);
         }
 
+        // Load empty EagerLoadedLists into any parents which have no child records
+        $emptyRelations = array_diff($parentIDs, array_keys($eagerLoadedLists));
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
        $missingParentIDs = array_diff($parentIDs, array_keys($eagerLoadedLists));
```

File: src/ORM/DataList.php [hunk 8]
```diff
@@ -1234,12 +1261,24 @@ private function fetchEagerLoadHasMany(
...
+        $this->fillEmptyEagerLoadedRelations(
+            $parents,
+            $emptyRelations,
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
            $missingParentIDs,
```

File: src/ORM/DataList.php [hunk 9]
```diff
@@ -1307,14 +1348,27 @@ private function fetchEagerLoadManyMany(
...
                 $eagerLoadedLists[$parentID] = $eagerLoadedList;
-                $this->addEagerLoadedDataToParent($parents, $parentID, $relationChain, $relationName, $eagerLoadedList, 'many_many');
+                $this->addEagerLoadedDataToParent($parents, $parentID, $relationChain, $relationName, $eagerLoadedList, $relationType);
             }
             // Add this row to the list
             $eagerLoadedList->addRow($relationItem);
         }
 
+        // Load empty EagerLoadedLists into any parents which have no child records
+        $emptyRelations = array_diff($parentIDs, array_keys($eagerLoadedLists));
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
        $missingParentIDs = array_diff($parentIDs, array_keys($eagerLoadedLists));
```

File: src/ORM/DataList.php [hunk 10]
```diff
@@ -1307,14 +1348,27 @@ private function fetchEagerLoadManyMany(
...
+        $this->fillEmptyEagerLoadedRelations(
+            $parents,
+            $emptyRelations,
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
            $missingParentIDs,
```

File: src/ORM/DataList.php [hunk 11]
```diff
@@ -1367,6 +1421,46 @@ private function addEagerLoadedDataToParent(
         }
     }
 
+    private function fillEmptyEagerLoadedRelations(
+        Query|array $parents,
+        array $parentIDs,
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
        array $missingParentIDs,
```

File: src/ORM/DataList.php [hunk 12]
```diff
@@ -1367,6 +1421,46 @@ private function addEagerLoadedDataToParent(
...
+        string $relationChain,
+        string $relationName,
+        string $relationType,
+        string $relationDataClass,
+        ?string $relationListClass = null,
+        ?array $component = null
+    ): void {
+        foreach ($parentIDs as $id) {
```
Comment (emteknetnz) [inline src/ORM/DataList.php]:
```suggestion
        foreach ($missingParentIDs as $id) {
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Just need to update variable names so things are consistent and easier to understand

File: src/ORM/DataList.php (hunk 12 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: done

File: src/ORM/DataList.php (hunk 11 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: These aren't missing.... they're known IDs that don't have anything in their relation. But not worth discussing IMO so I've made the change.

File: src/ORM/DataList.php (hunk 10 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Done

File: src/ORM/DataList.php (hunk 9 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Done

File: src/ORM/DataList.php (hunk 8 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Done

File: src/ORM/DataList.php (hunk 7 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Done

File: src/ORM/DataList.php (hunk 6 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Done

File: src/ORM/DataList.php (hunk 5 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Both `emptyRelations` and `missingParentIDs` are correct.... but not worth discussing, I've made the change.

File: src/ORM/DataList.php (hunk 4 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Done

File: src/ORM/DataList.php (hunk 3 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Done

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**